### PR TITLE
Makes search action cancelable by task management API

### PR DESCRIPTION
**Pull request by imotov**
Long running searches now can be cancelled using standard task cancellation mechanism.

### CODE REVIEW

**Comment by nik9000** [comment]
@imotov, want me to review it?

**Comment by imotov** [comment]
@nik9000 sure - you from Task Management perspective and we will need somebody closer to Lucene ideally.

File: core/src/main/java/org/elasticsearch/action/search/SearchTask.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Task storing information about a currently running search request.
+ */
+public class SearchTask extends CancellableTask {
+
+    public SearchTask(long id, String type, String action, String description, TaskId parentTaskId) {
+        super(id, type, action, description, parentTaskId);
+    }
+
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/search/SearchTask.java]: It'd be cool to be able to list the phase and/or which shards you are waiting for. You could put all kinds of cool stuff in here one day! But for now this seems like the right thing to do.

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 2]
```diff
@@ -227,10 +229,11 @@ protected void doClose() {
         keepAliveReaper.cancel();
     }
 
-    public DfsSearchResult executeDfsPhase(ShardSearchRequest request) throws IOException {
+    public DfsSearchResult executeDfsPhase(ShardSearchRequest request, SearchTask task) throws IOException {
         final SearchContext context = createAndPutContext(request);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: Maybe `createContext` should take the `SearchTask` as an argument? That'd make it very difficult to forget to set it.

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 3]
```diff
@@ -290,11 +294,12 @@ public QuerySearchResultProvider executeQueryPhase(ShardSearchRequest request) t
         }
     }
 
-    public ScrollQuerySearchResult executeQueryPhase(InternalScrollSearchRequest request) {
+    public ScrollQuerySearchResult executeQueryPhase(InternalScrollSearchRequest request, SearchTask task) {
         final SearchContext context = findContext(request.id());
         SearchOperationListener operationListener = context.indexShard().getSearchOperationListener();
         context.incRef();
         try {
+            context.setTask(task);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: Oooh! Yeah! That makes sense - scrolls contexts will outlast that task that spawns them and need to get a new task.

I haven't read the whole PR yet, but I wonder if we should clear task from the context between requests? If we don't then we maintain a reference to it after the task manager and that seems a bit weird. Changes to it wouldn't be useful to anyone.

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 4]
```diff
@@ -633,6 +643,7 @@ private void cleanContext(SearchContext context) {
         try {
             assert context == SearchContext.current();
             context.clearReleasables(Lifetime.PHASE);
+            context.setTask(null);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: And here is the clear. Cool.

File: core/src/main/java/org/elasticsearch/action/search/SearchTask.java (hunk 1 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/action/search/SearchTask.java]: Yes, that's the plan, but I didn't want to clump everything into a single PR.

File: core/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java [hunk 5]
```diff
@@ -58,6 +59,9 @@ public void execute(SearchContext context) {
             TermStatistics[] termStatistics = new TermStatistics[terms.length];
             IndexReaderContext indexReaderContext = context.searcher().getTopReaderContext();
             for (int i = 0; i < terms.length; i++) {
+                if(context.isCancelled()) {
+                    throw new SearchContextException(context, "Cancelled");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java]: This one is upper case and the one below is lower case.

File: core/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 6]
```diff
@@ -114,6 +116,18 @@ public ParseFieldMatcher parseFieldMatcher() {
         return parseFieldMatcher;
     }
 
+    public void setTask(SearchTask task) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: It'd be nice to have javadoc for these, especially given that task changes for scrolls.

File: core/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 7]
```diff
@@ -102,6 +103,7 @@ public static SearchContext current() {
     private Map<Lifetime, List<Releasable>> clearables = null;
     private final AtomicBoolean closed = new AtomicBoolean(false);
     private InnerHitsContext innerHitsContext;
+    private SearchTask task;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: It is probably worth some javadoc explaining why this isn't `final` and why it is ok that it isn't `volatile`.

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private class CancellableLeafCollector extends FilterLeafCollector {
+        private CancellableLeafCollector(LeafCollector in) {
+            super(in);
+        }
+
+        @Override
+        public void collect(int doc) throws IOException {
+            if (searchContext.isCancelled()) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: It is worth benchmarking this because it adds a volatile read. If it is too heavy we can make the check less frequent I imagine.

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 9]
```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
                 }
             }
 
+            if (collector != null && searchContext.getTask() != null) {
+                final Collector child = collector;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: A null task here is pretty weird, right?

**Comment by nik9000** [comment]
@jpountz, would you like to have a look at this from a Lucene perspective?

File: core/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 10]
```diff
@@ -447,6 +448,27 @@ public void removeConnectionListener(TransportConnectionListener listener) {
...
+                                                               final TransportResponseHandler<T> handler) {
+        request.setParentTask(localNode.getId(), parentTask.getId());
+        try {
+            taskManager.registerChildTask(parentTask, node.getId());
+        } catch (IllegalStateException ex) {
+            // The parent task is already cancelled - just fail the request
+            handler.handleException(new TransportException(ex));
+            return;
+        }
+        sendRequest(node, action, request, options, handler);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/transport/TransportService.java]: nit: I'd prefer that we move the sendRequest call to the try block and remove the return from the catch block

File: core/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 11]
```diff
@@ -114,6 +120,18 @@ public ParseFieldMatcher parseFieldMatcher() {
...
+    public void setTask(SearchTask task) {
+        this.task = task;
+    }
+
+    public SearchTask getTask() {
+        return task;
+    }
+
+    public boolean isCancelled() {
+        return task != null && task.isCancelled();
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: I don't think task should ever be null at that point? Should we just assert that `task` is not null?

File: core/src/main/java/org/elasticsearch/search/SearchService.java (hunk 2 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: +1

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 12]
```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
                 }
             }
 
+            if (collector != null) {
+                final Collector child = collector;
+                collector = new CancellableCollector(searchContext, collector);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: This part can cause significant slow downs. There may be ways that we can make it better, eg. by doing it at the bulk scorer level so that we can perform the check less often. Can we remove it for now and work on making this part cancellable in another PR. Or at least make it an opt-in for now with a setting?

**Review by jpountz — CHANGES_REQUESTED** [review]
In general it looks good to me, however I am concerned that wrapping the collector can cause significant slow downs so I'd like to either remove it or make it an opt-in for now and see what other options we have. For instance I'm wondering that we could make the slow down more acceptable by checking whether tasks are cancelled at the bulk scorer level so that we can perform the check less often.

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java (hunk 12 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: @jpountz  what do you think about this 662e1f5e3a2e9db89511749fe104e922edbb86ad?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: I'm not a fan of the name (but I don't have better ideas) but this looks good to me.

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import java.io.IOException;
+
+/**
+ * Collector that checks if the task it is executed under is cancelled
+ */
+public class CancellableCollector extends FilterCollector {
+    private final SearchContext searchContext;
+    private final boolean leafLevel;
+
+    public CancellableCollector(SearchContext searchContext, Collector in) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: Can we avoid taking a SearchContext as it makes this class hard to unit test, maybe we could eg. replace the search context with a `BooleanSupplier` and use a method reference? Then can you add some basic unit tests to this class?

File: core/src/main/java/org/elasticsearch/search/profile/query/CollectorResult.java [hunk 14]
```diff
@@ -45,6 +45,7 @@
     public static final String REASON_SEARCH_MIN_SCORE = "search_min_score";
     public static final String REASON_SEARCH_MULTI = "search_multi";
     public static final String REASON_SEARCH_TIMEOUT = "search_timeout";
+    public static final String REASON_SEARCH_CANCELED = "search_canceled";
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/profile/query/CollectorResult.java]: should we be consistent with the number of `l`? there is only one here while you put 2 in `cancellation`? (feel free to ignore, it could easily a special case in english that I don't know about!)

File: core/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 15]
```diff
@@ -220,6 +238,11 @@ public InnerHitsContext innerHits() {
 
     public abstract void terminateAfter(int terminateAfter);
 
+    // Indicates if the current index should perform frequent low level search cancellation check
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: Can you make it an actual javadoc and explain the trade-off?

**Review by jpountz — COMMENTED** [review]
I left a few more comments about the search part of the PR.

File: core/src/main/java/org/elasticsearch/search/SearchService.java (hunk 2 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: See @nik9000's comment bellow, this doesn't seem to be a good way to handle it since the task can change over time.

**Comment by imotov** [comment]
@jpountz I have implemented the changes that you have requested and rebased it against the current master since it quite a few things has changed. Would you mind taking another look when you have a chance?

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Collector that checks if the task it is executed under is cancelled
+ */
+public class CancellableCollector extends FilterCollector {
+    private final Provider<Boolean> cancelled;
+    private final boolean leafLevel;
+
+    /**
+     * Constractor
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: construction contractor? :)

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 17]
```diff
@@ -546,6 +567,7 @@ final SearchContext createContext(ShardSearchRequest request, @Nullable Engine.S
                 keepAlive = request.scroll().keepAlive().millis();
             }
             context.keepAlive(keepAlive);
+            context.lowLevelCancellation(lowLevelCancellation);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: why do we only do it in the fetch phase?

**Review by jpountz — COMMENTED** [review]
I left a question.

File: core/src/main/java/org/elasticsearch/search/SearchService.java (hunk 17 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: sorry I just realized this is the createContext method and not executeFetchPhase as I initially thought.

then I'm wondering if we could have the lowLevelCancellation only on DefaultSearchContext rather than the SearchContext base class?

Comment (imotov) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: done

**Comment by jpountz** [comment]
The search part looks good to me.

File: core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+/**
+ * Transport request handlers that is using task context
+ */
+public abstract class TaskAwareTransportRequestHandler<T extends TransportRequest> implements TransportRequestHandler<T>  {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java]: I guess you could make this a final class that takes a functional interface as its only parameter and runs it and you could still use lambdas for it. My instincts are that that is marginally better from a design standpoint (composition vs inheritance) and a little easier to read but I'm not sure.

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 19]
```diff
@@ -107,6 +109,8 @@
         Setting.positiveTimeSetting("search.default_keep_alive", timeValueMinutes(5), Property.NodeScope);
     public static final Setting<TimeValue> KEEPALIVE_INTERVAL_SETTING =
         Setting.positiveTimeSetting("search.keep_alive_interval", timeValueMinutes(1), Property.NodeScope);
+    public static final Setting<Boolean> LOW_LEVEL_CANCELLATION_SETTING =
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: Can haz javadoc?

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Collector that checks if the task it is executed under is cancelled
+ */
+public class CancellableCollector extends FilterCollector {
+    private final Provider<Boolean> cancelled;
+    private final boolean leafLevel;
+
+    /**
+     * Constructor
+     * @param cancelled supplier of the cancellation flag
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: Maybe mention that this amounts to a volatile read.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.index.query.QueryBuilders.scriptQuery;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+/**
+ */
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Can you remove the empty javdoc? We are going to fail the build on those at some point....

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE)
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Can you explain why this is needed in a comment? I'm never sure when to use this and when not to so I figure it'd be nice to have it explained.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder().put(SearchService.LOW_LEVEL_CANCELLATION_SETTING.getKey(), true).build();
+    }
+
+    private void indexTestData() {
+        for (int i = 0; i < 100; i++) {
+            client().prepareIndex("test", "type", Integer.toString(i)).setSource("field", "value").get();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Is this any quicker if you use bulks? I tend to do that out of habit.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+            if (i % 10 == 0) {
+                // Make sure we have a few segments
+                client().admin().indices().prepareRefresh("test").get();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Probably worth asserting that they return success. I've never seen them not do so but if they ever start then this test won't have many segments and will fail in strange ways to anyone trying to debug it.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    private void cancelSearch(String action) {
+        ListTasksResponse listTasksResponse = client().admin().cluster().prepareListTasks().setActions(action).get();
+        assertThat(listTasksResponse.getTasks().size(), equalTo(1));
+        TaskInfo searchTask = listTasksResponse.getTasks().get(0);
+
+        logger.info("Cancelling search");
+        CancelTasksResponse cancelTasksResponse = client().admin().cluster().prepareCancelTasks().setTaskId(searchTask.getTaskId()).get();
+        assertThat(cancelTasksResponse.getTasks().size(), equalTo(1));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: I find that `equalTo` is almost always a bad choice. In this case I think `assertThat(cancelTaskResponse.getTask(), hasSize(1));` will do the same thing but have much better error reporting. That way you get to see all the tasks when there are too many. Same for the above assertion.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+        assertThat(cancelTasksResponse.getTasks().get(0).getTaskId(), equalTo(searchTask.getTaskId()));
+    }
+
+    private void ensureSearchWasCancelled(ListenableActionFuture<SearchResponse> searchResponse) {
+        try {
+            SearchResponse response = searchResponse.actionGet();
+            logger.info("Search response {}", response);
+            if (response.getFailedShards() == 0) {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: `assertEquals("At least one shard should have failed", 0, response.getFailedShards());`?

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+        ListenableActionFuture<SearchResponse> scrollResponse = client().prepareSearchScroll(searchResponse.getScrollId()).execute();
+
+        awaitForBlock(plugins);
+        cancelSearch(SearchScrollAction.NAME);
+        disableBlocks(plugins);
+
+        ensureSearchWasCancelled(scrollResponse);
+
+        // Occasionally, cancellation might not hit all shards, leaving the scroll id hanging on these shards
+        client().prepareClearScroll().addScrollId(scrollId).get();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Does this happen because the scroll doesn't hit docs in that segment so it can't be blocked?

Should canceling a scroll request nuke the scroll id? If you tried to pump it again would it be busted? Maybe something for a followup though.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+public class SearchCancellationIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singleton(ScriptedBlockPlugin.class);
+    }
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder().put(SearchService.LOW_LEVEL_CANCELLATION_SETTING.getKey(), true).build();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: I wonder if it is worth only turning this on for the tests that needs it so it is obvious which ones do?

File: core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+
+    public void testLowLevelCancellableCollector() throws IOException {
+        TotalHitCountCollector collector = new TotalHitCountCollector();
+        AtomicBoolean cancelled = new AtomicBoolean();
+        CancellableCollector cancellableCollector = new CancellableCollector(cancelled::get, true, collector);
+        final LeafCollector leafCollector = cancellableCollector.getLeafCollector(reader.leaves().get(0));
+        leafCollector.collect(0);
+        cancelled.set(true);
+        try {
+            leafCollector.collect(1);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java]: `expectThrows`?

File: core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+    public void testCancellableCollectorBench() throws IOException {
+        TotalHitCountCollector collector = new TotalHitCountCollector();
+        AtomicBoolean cancelled = new AtomicBoolean();
+        CancellableCollector cancellableCollector = new CancellableCollector(cancelled::get, true, collector);
+        final LeafCollector leafCollector = cancellableCollector.getLeafCollector(reader.leaves().get(0));
+        long start = System.nanoTime();
+        for(int i=0; i<1000; i++) {
+            leafCollector.collect(0);
+        }
+        logger.info("took: {}", System.nanoTime() - start);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java]: If we keep this maybe we should rename it so it doesn't run by default? And javadoc it? It is fairly obvious what it is for but I expect it isn't worth doing it during a regular test run because no one is reading it....

File: docs/reference/search.asciidoc [hunk 31]
```diff
@@ -92,6 +92,18 @@ timeout. The setting key is `search.default_search_timeout` and can be
 set using the <<cluster-update-settings>> endpoints. Setting this value
 to `-1` resets the global search timeout to no timeout.
 
+[float]
+[[global-search-cancellation]]
+== Search Cancellation
+
+Long running searches can be cancelled using standard <<task-cancellation,task cancellation>>
```
Comment (nik9000) [inline docs/reference/search.asciidoc]: s/Long running search/Searchs/ ?

File: docs/reference/search.asciidoc [hunk 32]
```diff
@@ -92,6 +92,18 @@ timeout. The setting key is `search.default_search_timeout` and can be
...
+mechanism. By default, a running search only checks if it is cancelled or
+not on segment boundaries, therefore the cancellation can be delayed by large
+segments. The search cancellation responsiveness can be improved by setting
+the dynamic cluster-level setting `search.low_level_cancellation` to `true`.
```
Comment (nik9000) [inline docs/reference/search.asciidoc]: Mention that it only effects searches after the change to the setting.

Comment (nik9000) [inline docs/reference/search.asciidoc]: Are we thinking of automatically opting certain thing into this setting? Maybe we can have search benchmarks that compare?

File: docs/reference/search/profile.asciidoc [hunk 33]
```diff
@@ -150,7 +157,8 @@ This will yield the following result:
 // TESTRESPONSE[s/"build_scorer": 42602/"build_scorer": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.build_scorer/]
 // TESTRESPONSE[s/"create_weight": 89323/"create_weight": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.create_weight/]
 // TESTRESPONSE[s/"next_doc": 2852/"next_doc": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.next_doc/]
-// TESTRESPONSE[s/"time": "0.06989100000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.3043110000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.03227300000ms"/"time": $body.profile.shards.0.searches.0.collector.0.children.0.time/]
```
Comment (nik9000) [inline docs/reference/search/profile.asciidoc]: I'm sorry about this nightmare.....

Comment (nik9000) [inline docs/reference/search/profile.asciidoc]: I'd love some way to say "just match anything in this position that looks like X" so you don't have to describe the whole path....

File: docs/reference/search/profile.asciidoc [hunk 34]
```diff
@@ -390,21 +398,30 @@ Looking at the previous example:
...
 // TESTRESPONSE[s/]$/]}], "aggregations": []}]}}/]
-// TESTRESPONSE[s/"time": "0.06989100000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.3043110000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.03227300000ms"/"time": $body.profile.shards.0.searches.0.collector.0.children.0.time/]
 
-We see a single collector named `SimpleTopScoreDocCollector`.  This is the default "scoring and sorting" Collector
-used by Elasticsearch.  The `"reason"` field attempts to give a plain english description of the class name.  The
+We see a single collector named `SimpleTopScoreDocCollector` wrapped into `CancellableCollector`. `SimpleTopScoreDocCollector` is the default "scoring and sorting"
+Collector used by Elasticsearch.  The `"reason"` field attempts to give a plain english description of the class name.  The
```
Comment (nik9000) [inline docs/reference/search/profile.asciidoc]: `Collector`?

**Review by nik9000 — APPROVED** [review]
I left 19 review comments, mostly just comments and a few nits. Please make whatever changes you think are appropriate and merge when ready.

This is a huge start for canceling searches but I think it is worth expanding on the PR's description so folks know exactly what they are getting. Even with it's limitations I think this should be a release highlight for whatever release gets it. 5.1 I think....

**Review by jpountz — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java (hunk 18 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java]: This sounds like an interesting idea, but I think we should do it on the TransportRequestHandler then and should, probably, be another PR.

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java (hunk 20 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: It's supplier implementation detail, isn't it? It seems to be wrong place to say something like this. However, the information about how often it will be called can be useful here. I will add that.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java (hunk 22 above)
Comment (imotov) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Because I want nodes with specific settings and I don't want other tests to run with these settings.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java (hunk 28 above)
Comment (imotov) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Good catch! That was supposed to `randomBoolean()`. I replaced it with `true` while I was chasing a specific failure and forgot to flip it back on. I am unit testing that lowLevel is getting applied properly, so all tests should pass with either setting.

File: core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java (hunk 18 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java]: ++

File: core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java (hunk 20 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java]: Something like "This class assumes that the supplier is fast, with performance on the order of a volatile read." would give a lot of context to the decisions around how to use the Supplier.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java (hunk 28 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Awesome.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java (hunk 22 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: Then could you do something like:
```
@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE) // Changes settings
```

That'd make it clear why it is needed so folks don't go copying it into their tests without knowing why... 👼

File: docs/reference/search/profile.asciidoc (hunk 34 above)
Comment (nik9000) [inline docs/reference/search/profile.asciidoc]: Like, with the backticks.

File: core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java (hunk 30 above)
Comment (imotov) [inline core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java]: Leftovers, deleted.
